web/caret_browsing.js

```diff
@@ -0,0 +1,321 @@
+   * parameter).
+   * @param {boolean} select
+   * @param {Selection} selection
+   * @param {Element} newLineElement
+   * @param {DOMRect} newLineElementRect
+   * @param {number} caretX
+   */
+  #setCaretPosition(
+    select,
+    selection,
+    newLineElement,
+    newLineElementRect,
+    caretX
+  ) {
+    if (this.#isVisible(newLineElementRect)) {
+      this.#setCaretPositionHelper(
+        selection,
+        caretX,
+        select,
+        newLineElement,
+        newLineElementRect
+      );
+      return;
+    }
+    this.#mainContainer.addEventListener(
+      "scrollend",
+      this.#setCaretPositionHelper.bind(
+        this,
+        selection,
+        caretX,
+        select,
+        newLineElement,
+        null
+      ),
+      { once: true }
+    );
+    newLineElement.scrollIntoView();
+  }
+
+  /**
+   * Get the node on the next page.
+   * @param {Element} textLayer
+   * @param {boolean} isUp
+   * @returns {Node}
+   */
+  #getNodeOnNextPage(textLayer, isUp) {
+    while (true) {
+      const page = textLayer.closest(".page");
+      const pageNumber = parseInt(page.getAttribute("data-page-number"));
+      const nextPage = isUp ? pageNumber - 1 : pageNumber + 1;
+      textLayer = this.#viewerContainer.querySelector(
+        `.page[data-page-number="${nextPage}"] .textLayer`
+      );
+      if (!textLayer) {
+        return null;
+      }
+      const walker = document.createTreeWalker(textLayer, NodeFilter.SHOW_TEXT);
+      const node = isUp ? walker.lastChild() : walker.firstChild();
+      if (node) {
+        return node;
+      }
+    }
+  }
+
+  /**
+   * Move the caret in the given direction.
+   * @param {boolean} isUp
+   * @param {boolean} select
+   */
+  moveCaret(isUp, select) {
+    const selection = document.getSelection();
+    if (selection.rangeCount === 0) {
+      return;
+    }
+    const { focusNode } = selection;
+    const focusElement =
+      focusNode.nodeType !== Node.ELEMENT_NODE
+        ? focusNode.parentElement
+        : focusNode;
+    const root = focusElement.closest(".textLayer");
+    if (!root) {
+      return;
+    }
+    const walker = document.createTreeWalker(root, NodeFilter.SHOW_TEXT);
+    walker.currentNode = focusNode;
+
+    // Move to the next element which is not on the same line as the focus
+    // element.
+    const focusRect = focusElement.getBoundingClientRect();
+    let newLineElement = null;
+    const nodeIterator = (
+      isUp ? walker.previousSibling : walker.nextSibling
+    ).bind(walker);
+    while (nodeIterator()) {
+      const element = walker.currentNode.parentElement;
+      if (!this.#isOnSameLine(focusRect, element.getBoundingClientRect())) {
+        newLineElement = element;
+        break;
+      }
+    }
+
+    if (!newLineElement) {
+      // Need to find the next line on the next page.
+      const node = this.#getNodeOnNextPage(root, isUp);
+      if (node) {
```

Comment:
Can we reduce the overall indentation a bit below with an early return?
```suggestion
      if (!node) {
        return;
      }
```